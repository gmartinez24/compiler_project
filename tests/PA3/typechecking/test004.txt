// Have functions return the incorrect type

main

int lower, upper, totalOdd;

function isOdd (int x) : bool
{
    return x % 2; // Incorrect
};

function numOddsRange (int a, int b) : int
{
    int count;

    if (a > b) then
        return 0; // Should be fine
    fi;

    count = 0;

    while (a < b) do
        if (call isOdd(a)) then // Should be fine even though return is wrong
            count++;
        fi;
        a++;
    od;

    return count > 0; // Not fine
};

{
    lower = call readInt();
    upper = call readInt();

    totalOdd = call numOddsRange(lower, upper);
    call printInt(totalOdd);
    call println();
}.