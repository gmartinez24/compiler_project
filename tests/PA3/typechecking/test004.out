Error type-checking file.
TypeError(9,5)[Function isOdd returns int instead of bool.]
TypeError(29,5)[Function numOddsRange returns bool instead of int.]

